tsl/src/continuous_aggs/refresh.c

```diff
@@ -372,6 +386,19 @@ continuous_agg_refresh(PG_FUNCTION_ARGS)
 				 errmsg("invalid refresh window"),
 				 errhint("The start of the window must be before the end.")));
 
+	LockRelationOid(catalog_get_table_id(catalog, CONTINUOUS_AGGS_INVALIDATION_THRESHOLD),
+					AccessExclusiveLock);
+
+	if (continuous_agg_invalidation_threshold_set(cagg->data.raw_hypertable_id, refresh_window.end))
```

Comment:
TBH, I don't understand the numerous locks taken in the materializer code. Would appreciate some explanation and guidance. In particular, I don't understand the session locks that last across the two transactions. There's a comment that says "SessionLock the objects we are using for this materialization to ensure they're not alter in a way which would invalidate our work." The objects locked are materialized and raw hypertables, the various views, etc. But I can't make sense of this since the first transaction only moves the invalidations from hyper to cagg log and (potentially) updates the invalidation threshold. Thus the work in the first transaction doesn't really depend on these objects (beyond the individual thresholds and invalidation entries). 

The only thing  that I can see that these session-locks protect against is dropping the objects between the first and the second transaction. But since the work done is to process invalidations for _all_ hypertables, then why aren't _all_ hypertables referenced in the log locked, including all their caggs and dependent objects?

Thus, the first transaction seems to be pure bookkeeping and I don't see the harm in simply failing the second transaction if the objects no longer exist, like you would normally do if, e.g., a table was dropped just as someone was about to query it. In any case, if we care about not having objects removed under our feet, I think it is fine to add these session locks later. They seem more to be there to avoid unexpected error for the user--not for correctness. But maybe I am missing something?

It seems also that many locks are taken just before a scan that then takes the same locks. If the concern is that the scan will release the lock before transaction end, it is now possible to specify that locks are kept after the scan. This means that many locks that are taken explicitly, then immediately taken again in the scan, are no longer needed since the scan takes the appropriate lock. I've never before seen this pattern of taking the same lock twice, neither in most of our code nor in PostgreSQL itself.

Reply:
Yes, they are being locked to prevent objects from being removed from under us. If a DROP or ALTER is in progress , you would want the materialization to wait and fail. And vice-versa. 
Yes, the explicit lock + catalog lock is due to the fact that the catalog code also acquires the lock and releases it. Having the locks in one place helps reason about it as well.

Reply:
Since this PR is just about setting the invalidation threshold, does it sound reasonable that we review the need for the session locks and possible additions separately (for later PR)?

Reply:
Yes, would be good to make a pass over the locking after all the functionality is in place.